File: openags/research/messaging/telegram.py
"""Telegram notification channel via Bot API."""

from __future__ import annotations

import logging

import httpx

from openags.models import TelegramConfig

logger = logging.getLogger(__name__)

_LEVEL_EMOJI = {
    "info": "ℹ️",
    "warning": "⚠️",
    "error": "❌",
    "success": "✅",
}


class TelegramChannel:
    """Send notifications to a Telegram chat via Bot API.

    Requires a bot token and chat_id. Create a bot via @BotFather.
    """

    def __init__(self, config: TelegramConfig) -> None:
        self._token = config.token.get_secret_value()
        self._chat_id = config.chat_id
        self._base_url = f"https://api.telegram.org/bot{self._token}"

    @property
    def channel_id(self) -> str:
        return "telegram"

    async def send_text(self, text: str) -> bool:
        return await self._send(text, parse_mode=None)

    async def send_markdown(self, markdown: str) -> bool:
        return await self._send(markdown, parse_mode="MarkdownV2")

    async def send_notification(
        self,
        title: str,
        body: str,
        level: str = "info",
    ) -> bool:
        emoji = _LEVEL_EMOJI.get(level, "ℹ️")
        text = f"{emoji} *{title}*\n\n{body}"
        return await self._send(text, parse_mode="MarkdownV2")

    async def health_check(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=10) as client:
                r = await client.get(f"{self._base_url}/getMe")
                return r.status_code == 200 and r.json().get("ok", False)
        except Exception as e:
            logger.warning("Telegram health check failed: %s", e)
            return False

    async def _send(self, text: str, parse_mode: str | None = None) -> bool:
        """Send a message via the Telegram Bot API."""
        payload: dict[str, str] = {
            "chat_id": self._chat_id,
            "text": text,
        }
        if parse_mode:
            payload["parse_mode"] = parse_mode

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(f"{self._base_url}/sendMessage", json=payload)
                if r.status_code == 200 and r.json().get("ok"):
                    return True
                logger.warning("Telegram send failed: %s", r.text)
                return False
        except Exception as e:
            logger.error("Telegram send error: %s", e)
            return False

File: packages/app/src/messaging/telegram.ts
/**
 * Telegram Bot Integration
 *
 * Send messages and receive updates via Telegram Bot API.
 */

export interface TelegramConfig {
  botToken: string
  /** Default chat ID for notifications */
  chatId?: string | number
}

export interface TelegramMessage {
  chat_id: string | number
  text: string
  parse_mode?: 'MarkdownV2' | 'HTML' | 'Markdown'
  disable_notification?: boolean
  reply_to_message_id?: number
}

export interface TelegramUpdate {
  update_id: number
  message?: {
    message_id: number
    from?: {
      id: number
      username?: string
      first_name?: string
    }
    chat: {
      id: number
      type: string
      title?: string
    }
    date: number
    text?: string
  }
}

const TELEGRAM_API = 'https://api.telegram.org'

export class TelegramBot {
  private config: TelegramConfig

  constructor(config: TelegramConfig) {
    this.config = config
  }

  /**
   * Send a text message.
   */
  async sendMessage(message: TelegramMessage): Promise<{ ok: boolean; message_id?: number }> {
    const url = `${TELEGRAM_API}/bot${this.config.botToken}/sendMessage`

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(message),
    })

    const result = await response.json() as { ok: boolean; result?: { message_id: number } }

    return {
      ok: result.ok,
      message_id: result.result?.message_id,
    }
  }

  /**
   * Send a notification to the default chat.
   */
  async notify(text: string, options?: { parseMode?: 'MarkdownV2' | 'HTML' }): Promise<boolean> {
    if (!this.config.chatId) {
      throw new Error('chatId not configured')
    }

    const result = await this.sendMessage({
      chat_id: this.config.chatId,
      text,
      parse_mode: options?.parseMode,
    })

    return result.ok
  }

  /**
   * Get recent updates (for polling).
   */
  async getUpdates(options?: { offset?: number; limit?: number; timeout?: number }): Promise<TelegramUpdate[]> {
    const url = new URL(`${TELEGRAM_API}/bot${this.config.botToken}/getUpdates`)

    if (options?.offset) url.searchParams.set('offset', String(options.offset))
    if (options?.limit) url.searchParams.set('limit', String(options.limit))
    if (options?.timeout) url.searchParams.set('timeout', String(options.timeout))

    const response = await fetch(url.toString())
    const result = await response.json() as { ok: boolean; result: TelegramUpdate[] }

    return result.ok ? result.result : []
  }

  /**
   * Set a webhook URL for receiving updates.
   */
  async setWebhook(url: string): Promise<boolean> {
    const apiUrl = `${TELEGRAM_API}/bot${this.config.botToken}/setWebhook`

    const response = await fetch(apiUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ url }),
    })

    const result = await response.json() as { ok: boolean }
    return result.ok
  }

  /**
   * Delete the webhook.
   */
  async deleteWebhook(): Promise<boolean> {
    const url = `${TELEGRAM_API}/bot${this.config.botToken}/deleteWebhook`

    const response = await fetch(url, { method: 'POST' })
    const result = await response.json() as { ok: boolean }
    return result.ok
  }

  /**
   * Get bot info.
   */
  async getMe(): Promise<{ id: number; username: string; first_name: string } | null> {
    const url = `${TELEGRAM_API}/bot${this.config.botToken}/getMe`

    const response = await fetch(url)
    const result = await response.json() as { ok: boolean; result?: { id: number; username: string; first_name: string } }

    return result.ok ? result.result! : null
  }

  /**
   * Send a document.
   */
  async sendDocument(
    chatId: string | number,
    document: Buffer | string,
    options?: { filename?: string; caption?: string }
  ): Promise<boolean> {
    const url = `${TELEGRAM_API}/bot${this.config.botToken}/sendDocument`

    const formData = new FormData()
    formData.append('chat_id', String(chatId))

    if (typeof document === 'string') {
      // URL to document
      formData.append('document', document)
    } else {
      // Buffer
      const blob = new Blob([document])
      formData.append('document', blob, options?.filename || 'document')
    }

    if (options?.caption) {
      formData.append('caption', options.caption)
    }

    const response = await fetch(url, {
      method: 'POST',
      body: formData,
    })

    const result = await response.json() as { ok: boolean }
    return result.ok
  }
}
